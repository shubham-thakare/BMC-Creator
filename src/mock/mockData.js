export const mockState = {
  app: {
    title: 'BMC-Creator',
    version: '1.0.0',
  },
  title: 'Business Model Canvas Title',
  version: '1.0.0',
  date: '13 Apr, 2020',
  noteTitle: 'Test title',
  noteDescription: 'Test description',
  noteBackgroundColor: '#f5f8fa',
  noteTextColor: '#000000',
  activeKey: 'keyPartners',
  isDialogOpen: true,
  keyPartners: {
    notes: [
      {
        key: new Date().getTime(),
        title: 'Test note 1',
        description: 'Test description',
        background: 'red',
        color: 'white',
      },
      {
        key: new Date().getTime() + 1,
        title: 'Test note 2',
        description: 'Test description',
        background: 'red',
        color: 'white',
      },
      {
        key: new Date().getTime() + 2,
        title: '',
        description: 'Test description',
        background: undefined,
        color: 'white',
      },
      {
        key: new Date().getTime() + 3,
        title: 'Test note 2',
        description: '',
        background: 'red',
        color: 'white',
      },
    ],
  },
  keyActivities: { notes: [] },
  keyResources: { notes: [] },
  valuePropositions: { notes: [] },
  customerRelationships: { notes: [] },
  channels: { notes: [] },
  customerSegments: { notes: [] },
  costStructure: { notes: [] },
  revenueStreams: { notes: [] },
};
