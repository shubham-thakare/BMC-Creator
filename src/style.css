@import "~@blueprintjs/core/lib/css/blueprint.css";
@import "~@blueprintjs/icons/lib/css/blueprint-icons.css";

:root {
  --app-background-color: #f5f8fa;
  --app-divider-color: #182026;
  --link-color: #182026;
  --link-hover-color: #818181;
  --white-color: #ffffff;
  --black-color: #000000;
  --red-color: #ff0000;
  --title-input-color: #000000;
  --title-input-placeholder-color: #8a9ba8;
  --placeholder-color: #8a9ba8;
  --title-input-font-size: 32px;
  --input-font-size: 14px;
  --navbar-link-font-size: 18px;
  --navbar-link-font-size-smaller: 14px;
  --input-font-family: -apple-system, "BlinkMacSystemFont", "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Open Sans", "Helvetica Neue", "Icons16", sans-serif;
  --spacing-scale-1: 8px;
  --elevation-1: 0px 0px 5px -2px;
  --elevation-2: -1px 2px 5px 0px rgba(0, 0, 0, 0.2), 2px 2px 2px 0 rgba(0, 0, 0, 0.19);
}

* {
  box-sizing: border-box;
}

body {
  font-family: -apple-system, "BlinkMacSystemFont", "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Open Sans", "Helvetica Neue", "Icons16", sans-serif;
  overflow-x: hidden;
  margin: 0;
  background: var(--app-background-color);
  user-select: none;
  -webkit-user-drag: none;
}

/* width */
::-webkit-scrollbar {
  width: 10px;
}

/* Track */
::-webkit-scrollbar-track {
  background: transparent;
}

/* Handle */
::-webkit-scrollbar-thumb {
  background: rgba(0, 0, 0, 0.25);
  -webkit-border-radius: 10px;
  border-radius: 10px;
}

/* Handle on hover */
::-webkit-scrollbar-thumb:hover {
  background: rgba(0, 0, 0, 0.35);
}

span:focus {
  outline: none;
}

.bp3-editable-text.bp3-editable-text-placeholder:hover, .bp3-editable-text.bp3-editable-text-editing:hover {
  box-shadow: 0 0 0 0 rgba(19,124,189,0), 0 0 0 0 rgba(19,124,189,0), inset 0 0 0 1px rgba(16,22,26,0.15);
}

.bp3-editable-text.bp3-editable-text-placeholder:before, .bp3-editable-text.bp3-editable-text-editing:before {
  box-shadow: none;
}

.bp3-editable-text.bp3-editable-text-placeholder, .bp3-editable-text.bp3-editable-text-editing {
  padding: 5px;
  border-radius: 5px;
}

button.bp3-button {
  outline: none;
}

.full-width {
  width: 100%;
}

.text-bold {
  font-weight: bold;
}

.text-italic {
  font-style: italic;
}

.no-animation {
  animation: none !important;
}