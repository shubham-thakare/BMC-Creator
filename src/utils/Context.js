import { createContext } from 'react';
import htmlToImage from 'html-to-image';
import wait from 'waait';
import cmsData from '../cms';

export const AppContext = createContext({});

const filter = (node) => {
  return node.tagName !== 'i';
};

const exportImage = async () => {
  hideNav();
  await wait(500);
  const node = document.getElementById('canvas-screen');
  closeNav();
  htmlToImage.toSvgDataURL(node, { filter: filter }).then((dataUrl) => {
    var link = document.createElement('a');
    link.download = 'my-image-name.svg';
    link.href = dataUrl;
    link.click();
  });
};

const hideNav = () => {
  document.getElementById('sidebar').style.left = '-220px';
  document.getElementById('canvas-screen').style.paddingLeft = '10px';
};

const openNav = () => {
  document.getElementById('sidebar').style.left = '0';
  document.getElementById('canvas-screen').style.paddingLeft = '230px';
  document.getElementById('open-button').style.display = 'none';
  document.getElementById('toggle-hide').style.display = 'block';
};

const closeNav = () => {
  document.getElementById('sidebar').style.left = '-170px';
  document.getElementById('canvas-screen').style.paddingLeft = '60px';
  document.getElementById('open-button').style.display = 'block';
  document.getElementById('toggle-hide').style.display = 'none';
};

export const contextWithoutState = {
  cmsData,
  exportImage,
  openNav,
  closeNav,
};
