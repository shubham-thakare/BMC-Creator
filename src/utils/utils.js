import { createContext } from 'react';
import htmlToImage from 'html-to-image';
import wait from 'waait';
import { UPDATE_STATE_FROM_FILE_DATA } from '../actions';

export const AppContext = createContext({});
export const ContextProvider = AppContext.Provider;
export const ContextConsumer = AppContext.Consumer;

const filter = (node) => {
  return node.tagName !== 'i';
};

export const exportImage = async (fileName) => {
  hideNav();
  await wait(500);
  const node = document.getElementById('canvas-screen');
  closeNav();
  htmlToImage.toSvgDataURL(node, { filter: filter }).then((dataUrl) => {
    const link = document.createElement('a');
    link.download = fileName + '.svg';
    link.href = dataUrl;
    link.click();
  });
};

export const hideNav = () => {
  try {
    document.getElementById('sidebar').style.left = '-220px';
    document.getElementById('canvas-screen').style.paddingLeft = '10px';
  } catch (ex) {}
};

export const openNav = () => {
  try {
    document.getElementById('sidebar').style.left = '0';
    document.getElementById('canvas-screen').style.paddingLeft = '230px';
    document.getElementById('open-button').style.display = 'none';
    document.getElementById('toggle-hide').style.display = 'block';
  } catch (ex) {}
};

export const closeNav = () => {
  try {
    document.getElementById('sidebar').style.left = '-170px';
    document.getElementById('canvas-screen').style.paddingLeft = '60px';
    document.getElementById('open-button').style.display = 'block';
    document.getElementById('toggle-hide').style.display = 'none';
  } catch (ex) {}
};

export const saveFile = (currentState, fileName) => {
  try {
    const pom = document.createElement('a');
    const blob = new Blob([JSON.stringify(currentState)], { type: 'text/xml' });
    pom.setAttribute('href', window.URL.createObjectURL(blob));
    pom.setAttribute('download', fileName + '.bmc');
    pom.dataset.downloadurl = ['text/xml', pom.download, pom.href].join(':');
    pom.draggable = true;
    pom.classList.add('dragout');
    pom.click();
  } catch (ex) {}
};

export const openFile = (dispatch) => {
  try {
    const inputElement = document.createElement('input');
    inputElement.type = 'file';
    inputElement.accept = '.bmc';
    inputElement.addEventListener('change', (event) => {
      const input = event.target;
      const reader = new FileReader();
      reader.onload = () => {
        const data = reader.result;
        dispatch({ action: UPDATE_STATE_FROM_FILE_DATA, payload: data });
      };
      reader.readAsText(input.files[0]);
    });
    inputElement.dispatchEvent(new MouseEvent('click'));
  } catch (ex) {}
};

export const trimString = (string) => {
  return string ? string.toString().trim() : '';
};

export const userVisitedBefore = () => {
  const userVisitedBefore = getCookie('userVisitedBefore');
  setCookie('userVisitedBefore', true, 365);

  return userVisitedBefore;
};

const setCookie = (cname, cvalue, exdays) => {
  const d = new Date();
  d.setTime(d.getTime() + exdays * 24 * 60 * 60 * 1000);
  const expires = 'expires=' + d.toUTCString();
  document.cookie = cname + '=' + cvalue + ';' + expires + ';path=/';
};

const getCookie = (cname) => {
  const name = cname + '=';
  const ca = document.cookie.split(';');
  for (let i = 0; i < ca.length; i++) {
    let c = ca[i];
    while (c.charAt(0) === ' ') {
      c = c.substring(1);
    }
    if (c.indexOf(name) === 0) {
      return c.substring(name.length, c.length);
    }
  }
  return '';
};
