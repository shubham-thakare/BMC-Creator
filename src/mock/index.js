import { mockState } from './mockData';

export { mockState };
