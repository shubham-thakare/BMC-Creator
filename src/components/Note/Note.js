import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import {
  Menu,
  Popover,
  Intent,
  PopoverPosition,
  PopoverInteractionKind,
} from '@blueprintjs/core';
import { NoteWrapper, NoteTitle, NoteDescription } from './styles/Note.style';
import { trimString, AppContext } from '../../utils';
import { DELETE_NOTE } from '../../actions';
import cmsData from '../../cms';

export const NoteMenu = ({ noteKey, activeKey }) => {
  const { dispatch } = useContext(AppContext);
  const { t_menu, b_delete } = cmsData;

  return (
    <Menu>
      <Menu.Divider title={t_menu} />
      <Menu.Item
        text={b_delete}
        icon="trash"
        intent={Intent.DANGER}
        onClick={() => {
          dispatch({ action: DELETE_NOTE, payload: { noteKey, activeKey } });
        }}
      />
    </Menu>
  );
};

const Note = ({ item, activeKey, ...props }) => {
  return (
    <>
      {(trimString(item.title).length > 0 ||
        trimString(item.description).length > 0) && (
        <Popover
          content={<NoteMenu noteKey={item.key} activeKey={activeKey} />}
          position={PopoverPosition.RIGHT_BOTTOM}
          interactionKind={PopoverInteractionKind.CLICK}
          targetClassName="full-width"
        >
          <NoteWrapper bgColor={item.background} color={item.color} {...props}>
            {item.title && item.title.length > 0 && (
              <NoteTitle>{item.title}</NoteTitle>
            )}
            {item.description && item.description.length > 0 && (
              <NoteDescription>{item.description}</NoteDescription>
            )}
          </NoteWrapper>
        </Popover>
      )}
    </>
  );
};

Note.propTypes = {
  item: PropTypes.object.isRequired,
  activeKey: PropTypes.string.isRequired,
};

export default Note;
