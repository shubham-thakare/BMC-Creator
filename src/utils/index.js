import { AppContext, contextWithoutState } from './Context';

export { AppContext, contextWithoutState };
