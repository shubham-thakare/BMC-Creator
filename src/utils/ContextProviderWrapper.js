import React, { useReducer } from 'react';
import PropTypes from 'prop-types';
import { ContextProvider } from '.';
import { reducer, appState } from '../reducer';
import { mockState } from '../mock';

export const ContextProviderWrapper = ({ children, useMockState }) => {
  const [state, dispatch] = useReducer(
    reducer,
    useMockState ? mockState : appState,
  );

  return (
    <ContextProvider value={{ state, dispatch }}>{children}</ContextProvider>
  );
};

ContextProviderWrapper.propTypes = {
  useMockState: PropTypes.bool,
};
