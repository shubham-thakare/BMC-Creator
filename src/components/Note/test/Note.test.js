import React from 'react';
import { mount } from 'enzyme';
import Note from '..';
import { ContextProviderWrapper } from '../../../utils';
import { mockState } from '../../../mock';
import { NoteTitle, NoteDescription, NoteWrapper } from '../styles/Note.style';
import { Popover, Menu } from '@blueprintjs/core';
import { NoteMenu } from '../Note';

let wrapper, wrapper2, contextMenu;

const props = {
  item: mockState.keyPartners.notes[0],
  activeKey: 'keyPartners',
};

const props2 = {
  item: mockState.keyPartners.notes[2],
  activeKey: 'keyPartners',
};

const contextMenuProps = {
  noteKey: 1587532687006,
  activeKey: 'keyPartners',
};

describe('Note Component', () => {
  beforeAll(() => {
    wrapper = mount(<Note {...props} />);

    wrapper2 = mount(<Note {...props2} />);

    contextMenu = mount(
      <ContextProviderWrapper useMockState>
        <NoteMenu {...contextMenuProps} />
      </ContextProviderWrapper>,
    );
  });

  it('To match snapshot of Note', async (done) => {
    expect(wrapper).toMatchSnapshot();
    done();
  });

  it('Context menu popover exists', async (done) => {
    expect(wrapper.find(Popover)).toHaveLength(1);
    done();
  });

  it('NoteWrapper exists', async (done) => {
    expect(wrapper.find(NoteWrapper)).toHaveLength(1);
    done();
  });

  it('Title exists', async (done) => {
    expect(wrapper.find(NoteTitle)).toHaveLength(1);
    done();
  });

  it('Description exists', async (done) => {
    expect(wrapper.find(NoteDescription)).toHaveLength(1);
    done();
  });

  it('Context menu exists', async (done) => {
    expect(contextMenu.find(Menu)).toHaveLength(1);
    done();
  });

  it('Simulate Edit operation', async (done) => {
    const add = contextMenu
      .find(Menu.Item)
      .find('li')
      .find('a')
      .at(0);
    add.simulate('click', {
      payload: { activeKey: 'keyPartners', noteKey: '1587532687006' },
    });
    done();
  });

  it('Simulate delete operation', async (done) => {
    const add = contextMenu
      .find(Menu.Item)
      .find('li')
      .find('a')
      .at(1);
    add.simulate('click');
    done();
  });

  it('Only description found in Note', async (done) => {
    expect(wrapper2.find(NoteDescription)).toHaveLength(1);
    done();
  });
});
