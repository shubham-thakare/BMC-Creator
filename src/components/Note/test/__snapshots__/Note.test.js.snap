// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note Component To match snapshot of Note 1`] = `
<Note
  activeKey="keyPartners"
  item={
    Object {
      "background": "red",
      "color": "white",
      "description": "Test description",
      "key": 1587394019949,
      "title": "Test note 1",
    }
  }
>
  <Blueprint3.Popover
    boundary="scrollParent"
    captureDismiss={false}
    content={
      <NoteMenu
        activeKey="keyPartners"
        noteKey={1587394019949}
      />
    }
    defaultIsOpen={false}
    disabled={false}
    fill={false}
    hasBackdrop={false}
    hoverCloseDelay={300}
    hoverOpenDelay={150}
    inheritDarkTheme={true}
    interactionKind="click"
    minimal={false}
    modifiers={Object {}}
    openOnTargetFocus={true}
    position="right-bottom"
    targetClassName="full-width"
    targetTagName="span"
    transitionDuration={300}
    usePortal={true}
    wrapperTagName="span"
  >
    <Manager>
      <span
        className="bp3-popover-wrapper"
      >
        <Reference
          innerRef={[Function]}
        >
          <InnerReference
            innerRef={[Function]}
            setReferenceNode={[Function]}
          >
            <Blueprint3.ResizeSensor
              onResize={[Function]}
            >
              <span
                className="bp3-popover-target full-width"
                onClick={[Function]}
              >
                <Notestyle__NoteWrapper
                  bgColor="red"
                  className="print-item"
                  color="white"
                  key=".0"
                >
                  <div
                    className="Notestyle__NoteWrapper-sc-2z8c2g-0 hYtmWh print-item"
                    color="white"
                  >
                    <Notestyle__NoteTitle>
                      <span
                        className="Notestyle__NoteTitle-sc-2z8c2g-1 gNJdcF"
                      >
                        Test note 1
                      </span>
                    </Notestyle__NoteTitle>
                    <Notestyle__NoteDescription>
                      <span
                        className="Notestyle__NoteDescription-sc-2z8c2g-2 ilDHHe"
                      >
                        Test description
                      </span>
                    </Notestyle__NoteDescription>
                  </div>
                </Notestyle__NoteWrapper>
              </span>
            </Blueprint3.ResizeSensor>
          </InnerReference>
        </Reference>
        <Blueprint3.Overlay
          autoFocus={true}
          backdropClassName="bp3-popover-backdrop"
          backdropProps={Object {}}
          canEscapeKeyClose={true}
          canOutsideClickClose={true}
          enforceFocus={true}
          hasBackdrop={false}
          isOpen={false}
          lazy={true}
          onClose={[Function]}
          transitionDuration={300}
          transitionName="bp3-popover"
          usePortal={true}
        />
      </span>
    </Manager>
  </Blueprint3.Popover>
</Note>
`;
